src/constraint_system/composer.rs

```diff
@@ -128,13 +128,23 @@ impl TurboComposer {
     }
 
     /// Evaluate the runtime value of a witness
-    #[allow(dead_code)]
-    pub(crate) fn evaluate_witness(&self, witness: &Witness) -> &BlsScalar {
-        &self.witnesses[&witness]
+    ///
+    /// # Safety
+    ///
+    /// Using the allocated scalar of a witness inside a gadget could produce
+    /// an unsound circuit.
```

Comment:
I think this could be reformulated.

The problem is that scalar evaluation on gadget construction isn't expected and could lead to different circuit representations for the same code.

If the user calls this function, he should be aware it is to perform operations outside the circuit that have no security value at all.

Reply:
`scalar evaluation on gadget construction isn't expected and could lead to different circuit representations for the same code` to me is having a circuit that is unsound (in the rust terminology). I'm happy to reformulate / adding more information. What about:

```
Witness evaluation inside a gadget isn't expected and could produce an unsound circuit (different circuit representation for the same code).

Calling this function performs operations outside the circuit.
```

If you want to add the part `that have no security value at all` I think we should expand the concept.

Personally I don't think we should mention how to prove and verify here,  because if get to read this comment but you don't know even how to prove and verify in plonk it means you just skipped all the rest of the library / tests.

Reply:
That sounds good